### ISSUE TITLE
Behaviour of Environment config source is not explicit

### ISSUE THREAD

**Issue by pilhuhn** [eclipse/microprofile-config#262]
The spec mentions that environment variables can be used as config sources.
It does not specify if those variables need to be converted to a different format.
From how I understood @Emily-Jiang an env variable of `MP_METRICS_TAGS` should be the same as a config key of `mp.metrics.tags` that was passed in via e.g. a system property.

cc @jmesnil

**Comment by jmesnil** [comment]
I don't think there is anything related to this in the 1.1 spec.
However we talked about it in the context of the JSR# 382 Config API (https://github.com/eclipse/ConfigJSR/issues/24)

We should provide the same behaviour for the mp-config 1.1 spec.

The JSR proposal is to substitute the '_' by '.' for environment config source (and only for this config source).
Your proposal goes further as it would also lower case the env variables.

@Emily-Jiang wydt?

Personnally, I would only substitute '_' by '.'
Nothing prevents to use lower-cased environment variables and I prefer to reduce the manipulation of property names to the minimum.

**Comment by pilhuhn** [comment]
@jmesnil not lowercasing would prevent me to transparently get a result for 

`@ConfigProperty(name="mp.metrics.tags")`

as the `name()` would not match for env-variables where the key has not been lowercased.

Or it would require me to uppercase the property and sysprops like

`java -DMP.METRICS.TAGS=foo` 

All the examples in the spec show lowercase names for properties.

**Comment by jmesnil** [comment]
what's preventing you to use `mp_metrics_tags` for the name of the env variable?

**Comment by pilhuhn** [comment]
that is not the unix way of naming env variables. See e.g. https://stackoverflow.com/a/673940/100957 or 

> Variables are case sensitive and capitalized by default

from http://tldp.org/LDP/Bash-Beginners-Guide/html/sect_03_02.html (section 3.2.2)

**Comment by jmesnil** [comment]
From the same source:

> you are free to use the names you want or to mix cases 

:)

I agree that usage is to use upper case it but nothing precludes you to lower case it.

```
export MP_METRICS_TAGS = whatever
export mp_metrics_tags = ${MP_METRICS_TAGS}
```

I'm my personal preference to avoid putting too much "magic" transformation in the microprofile-config specification to provide the least surprise. Substituting `_` by `.` is enough to support env variable.
Lower-casing the name adds another transformation that is not strictly required.

**Comment by jmesnil** [comment]
Lower-casing env variable would also prevent to use them to configure unnamed `@Config`property (as specified in the javadoc https://github.com/eclipse/microprofile-config/blob/master/api/src/main/java/org/eclipse/microprofile/config/inject/ConfigProperty.java#L102).

In the [TCK](https://github.com/eclipse/microprofile-config/blob/master/tck/src/main/java/org/eclipse/microprofile/config/tck/CDIPlainInjectionTest.java#L55) we have the property name `org.eclipse.microprofile.config.tck.CDIPlainInjectionTest.DefaultPropertyBean.configProperty` 

This could be configured with the env variable: `org_eclipse_microprofile_config_tck_CDIPlainInjectionTest_DefaultPropertyBean_configProperty`

But if we lower case env variables, we will no longer be able to use it to configure the unnamed `configProperty` field.

**Comment by Emily-Jiang** [comment]
This is tricky. We do want to make config name case sensitive. Otherwise, we will have many name clashes, as @jmesnil mentioned.
@pilhuhn you can do 

`@ConfigProperty(name="MP.METRICS.TAGS") `
or 
`@ConfigProperty(name="MP_METRICS_TAGS")`

Alternatively, in the metrics spec, you can introduce both MP_METRICS_TAGS and mp.metrics.tags. If the config is only used by the implementation, you can first search for mp.metrics.tags. If it does not exist, you then search for MP_METRICS_TAGS or MP.METRICS.TAGS. Basically, the new property overrides the old one.

**Comment by pilhuhn** [comment]
@Emily-Jiang I understand the issue with the clashes, but ENV variables are usually uppercase. That is a convention. 
And the properties are all lowercase in the examples.
Of course I can do the double checking, but at the end I lose one of the benefits of config to do the heavy lifting for me.

Java8 introduced the `@Repeatable` annotations - perhaps this could be a solution so that one could say

```
@ConfigProperty(name="MP_METRICS_TAGS")
@ConfigProperty(name="mp.metrics.tags")
String tags;
```

so that config could to the if-the-first-is-not-there-try-the-second thing under the covers?

**Comment by Emily-Jiang** [comment]
+1 @pilhuhn ! Sounds like this config is used by end users. Therefore, we need to put the simplicity and easy to use first. I totally agree your suggestion on letting the implementation to do the heavy lift. Making ConfigProperty to be Repeatable annotation should solve the problem. What do other people think?

**Comment by jmesnil** [comment]
+1 making the `@ConfigProperty` repeatable is a good idea.

The spec should also clarify how default value interact with repeated @ConfigProperty:
* @ConfigProperty are searched in the order of the annotations.
* The first @ConfigProperty that defines a value *or a default value* will be used.

In practice, this means that only the last @ConfigProperty should have a default value otherwise all subsequent ones would not be searched.

wdyt?

**Comment by Emily-Jiang** [comment]
+1 @jmesnil ! By the way, the repeatable annotation qualifier has dependency on CDI 2.0. Since MP has only CDI 1.2, this issue will have dependency on MP 2.0 release.

**Comment by jmesnil** [comment]
Another alternative could be to change the type of the @ConfigProperty's `name` to be `String[]` instead of `String`:

```
public @interface ConfigProperty {
    String[] name() default "";
}
```

With that change, we can specify either no name, a single name or an array of names (that would solve heiko's issue):

```
@Inject
@ConfigProperty(name={"MP_METRICS_TAGS", "mp.metrics.tags"})
String tags;
```

The advantage is that we keep a single @ConfigProperty annotation to configure a property. There is only 1 defaultValue, the names of the properties are ordered.

And afaict, we do not break existing user code, the code below is still valid:

```
@Inject
@ConfigProperty(name="my.double.property")
private double doubleProperty;

@Inject
// config property names is inferred from the injection point
@ConfigProperty
private int intProperty;
```

@Emily-Jiang wdyt?

**Comment by dmlloyd** [comment]
I believe this was fixed by cafee0a79fd36dfaa4f7b30bc77efa75eefa589b.  If I'm wrong please re-open!
